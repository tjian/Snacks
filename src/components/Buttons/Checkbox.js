import React, { PropTypes } from 'react'
import RadioCheckboxBase from '../../base/RadioCheckboxBase'
import svgCheckboxBase from '../../assets/checkboxBase.svg'
import svgCheckboxChecked from '../../assets/checkboxChecked.svg'

const ASSETS = {
  btnBkg: {
    base: svgCheckboxBase,
    checked: svgCheckboxChecked,
  }
}

const Checkbox = (props) => {
  return <RadioCheckboxBase {...props} assets={ASSETS} btnType='checkbox' />
}

Checkbox.propTypes = {
  aria          : PropTypes.shape({
    label         :PropTypes.string,
  }),
  children      : PropTypes.string,
  id            : PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  isSelected    : PropTypes.bool,
  name          : PropTypes.string,
  onClick       : PropTypes.func,
  styles        : PropTypes.shape({
    button        : PropTypes.object,
    label         : PropTypes.object,
    wrapEl        : PropTypes.object,
  }),
  value         : PropTypes.string,
  wrapEl        : PropTypes.string,
}

export default Checkbox
