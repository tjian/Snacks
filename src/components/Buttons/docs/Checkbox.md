This component outputs a checkbox with an optional label.

You can define whether the button starts in the pressed on unpressed state through the `isChecked` prop. In addition, you can change the checked/unchecked state at any moment after the mount state by redifining the prop `isChecked`.

### Label
Associating a label to every radio button is very important for accessibility purposes. There are two ways to do it. The easiest is to enclose the text with the component: `<Checkbox>My label</Checkbox>`. The other option is to associate a label yourself:

```js
<div style={{display: 'flex', flexDirection: 'column'}}>
  <div style={{display: 'flex', marginBottom: 20}}>
    <Checkbox id="checkbox1" isChecked />
    <label htmlFor="checkbox1" style={{marginLeft: 10}}>Label associated manually</label>
  </div>
  <Checkbox id="checkbox2">Label auto-generated</Checkbox>
</div>
```

### Styling
You can pass an object to the `styles` prop for styling the label, radio button and wrap element using Radium's structure:
```js
<Checkbox
  id="checkbox3"
  wrapEl='span'
  styles={{
    wrapEl: {padding: 10, border: '2px dotted lightblue'},
    button: {marginRight: 40},
    label: {color: 'salmon'},
  }}
>
  My Checkbox Button Label
</Checkbox>
```


### Callback
You can pass a callback through the `onClick` prop:

```js
function onClick(event, props) {
  alert(`This button is ${props.isSelected ? '' : 'not '}checked`)
}

<Checkbox id="checkbox4" onClick={onClick}>Click me!</Checkbox>
```
