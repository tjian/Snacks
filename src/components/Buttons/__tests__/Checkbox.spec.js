import Checkbox from '../Checkbox'
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import sinon from 'sinon'

describe('Checkbox', () => {
  it('renders the correct button type', () => {
    const tree = renderer
      .create(<Checkbox id={1} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders the correct selected state", () => {
    const testCases = [
      { id: 1, isSelected: false },
      { id: 2, isSelected: true },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<Checkbox {...props} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  
  it('generates a label if a child prop of text is supplied', () => {
    const tree = renderer
      .create(<Checkbox id={1}>Test label</Checkbox>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  });
  

  it('incorporates user styles if passed', () => {
    const testCases = [
      {
        id: 1,
        styles: {
          wrapEl: {background: 'pink'},
          button: {border: '2px solid purple' },
          label: {color: 'green'}
        },
      },
      {
        id: 2,
        styles: undefined
      },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<Checkbox {...props}>Test label</Checkbox>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('calls the user-supplied onClick function', () => {
    const onClick = sinon.spy()
    const testOnClick = () => testText
    const wrapper = mount(<Checkbox id={1} onClick={onClick} />)

    wrapper.find('input').simulate('click')
    expect(onClick.calledOnce).toBe(true)
  })
})
