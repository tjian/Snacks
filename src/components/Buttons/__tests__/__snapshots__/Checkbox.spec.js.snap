// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox generates a label if a child prop of text is supplied 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <img
      alt=""
      data-radium={true}
      src="test-file-stub"
      style={
        Object {
          "height": 22,
          "width": 22,
        }
      }
      tabIndex="-1"
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      id={1}
      onClick={[Function]}
      style={
        Object {
          "-moz-appearance": "none",
          "-ms-appearance": "none",
          "-webkit-appearance": "none",
          "appearance": "none",
          "height": 22,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="checkbox"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Checkbox incorporates user styles if passed 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "background": "pink",
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "border": "2px solid purple",
        "position": "relative",
      }
    }
  >
    <img
      alt=""
      data-radium={true}
      src="test-file-stub"
      style={
        Object {
          "height": 22,
          "width": 22,
        }
      }
      tabIndex="-1"
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      id={1}
      onClick={[Function]}
      style={
        Object {
          "-moz-appearance": "none",
          "-ms-appearance": "none",
          "-webkit-appearance": "none",
          "appearance": "none",
          "height": 22,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="checkbox"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "color": "green",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Checkbox incorporates user styles if passed 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <img
      alt=""
      data-radium={true}
      src="test-file-stub"
      style={
        Object {
          "height": 22,
          "width": 22,
        }
      }
      tabIndex="-1"
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      id={2}
      onClick={[Function]}
      style={
        Object {
          "-moz-appearance": "none",
          "-ms-appearance": "none",
          "-webkit-appearance": "none",
          "appearance": "none",
          "height": 22,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="checkbox"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={2}
    style={
      Object {
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Checkbox renders the correct button type 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <img
    alt=""
    data-radium={true}
    src="test-file-stub"
    style={
      Object {
        "height": 22,
        "width": 22,
      }
    }
    tabIndex="-1"
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    id={1}
    onClick={[Function]}
    style={
      Object {
        "-moz-appearance": "none",
        "-ms-appearance": "none",
        "-webkit-appearance": "none",
        "appearance": "none",
        "height": 22,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`Checkbox renders the correct selected state 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <img
    alt=""
    data-radium={true}
    src="test-file-stub"
    style={
      Object {
        "height": 22,
        "width": 22,
      }
    }
    tabIndex="-1"
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    id={1}
    onClick={[Function]}
    style={
      Object {
        "-moz-appearance": "none",
        "-ms-appearance": "none",
        "-webkit-appearance": "none",
        "appearance": "none",
        "height": 22,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`Checkbox renders the correct selected state 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <img
    alt=""
    data-radium={true}
    src="test-file-stub"
    style={
      Object {
        "height": 22,
        "width": 22,
      }
    }
    tabIndex="-1"
  />
  <input
    aria-label={undefined}
    checked={true}
    data-radium={true}
    id={2}
    onClick={[Function]}
    style={
      Object {
        "-moz-appearance": "none",
        "-ms-appearance": "none",
        "-webkit-appearance": "none",
        "appearance": "none",
        "height": 22,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;
